sys/include/ztimer.h

```diff
@@ -345,6 +345,20 @@ typedef struct {
      * @param   clock       ztimer clock to cancel a pending alarm, if any
      */
     void (*cancel)(ztimer_clock_t *clock);
+
+#if MODULE_ZTIMER_ONDEMAND || DOXYGEN
+    /**
+     * @brief   Starts the underlying timer
+     * @param   clock       ztimer clock to start
+     */
+    void (*start)(ztimer_clock_t *clock);
+
+    /**
+     * @brief   Stops the underlying timer
```

Comment:
this should not force the timer to be stopped but free the timer to be stopable when powermanagement requires that (that should be reflected in the name) (stoping and starting timer takes time sometimes we might not even know the amount (the compensation value might only be good for a average case) (e.g. if stating and stopping requires a crystal to get running or some mcu that have register sync cycles or need clock phases to align)

Reply:
Yeah, but in that case `nrf52` won't gain any power savings. Peripherals have to disabled explicitly. There's no special power management unit for `nrf52`, IIRC. Aside from that, `periph_timer`, ... don't have an API to indicate that the timer is allowed to stall silently.

Reply:
i just looked at it (seems like the nrf52 has its own PM task running in hardware), but there are other that do not, for nrf52 it might make sense to immediately stop, but for other mcu it might just free up a pmlevel.  mcu not have these things if you stop their timer (for pm) they might lose the status of that timer therefor the start has to reinitialize the timer

some cant be stopped  (pipico) (but if you do not need the timer you may not keep the osc running when it going to sleep)

Reply:
btw there is a possible race condition in the nrf52 hardware os that might throw of your timing if you immediately stop at 0 users (the manual warns about starting and stopping in the same PCLK16M (stop will win) ) so stopping should wait until you are quiet sure it will no be started soon.
for the sam you need to do what the nrf52 does for you in software ( disable unneed clocks) you might stop the timer but that will not safe much power.

Reply:
> mcu not have these things if you stop their timer (for pm) they might lose the status of that timer therefor the start has to reinitialize the timer

I expect `timer_stop()` not to completely deinit the peripheral. And I haven't observed this happening on any cpu I worked with.

> some cant be stopped (pipico) (but if you do not need the timer you may not keep the osc running when it going to sleep)

I this case the `timer_stop()` and `timer_start()` just interact with `pm_layered` and block resp. unblock the right pm state.

--

What is your suggestion how to improve this timer system? I'm still convinced that the direct interaction with `pm_layered` is a bad idea. I've collected many lost hours because of silly bugs caused by this implementation. Its pure pain in the butt.

Reply:
if `*_stop` does may not stop and `*_start` may not start why have it named start and stop (we need to write "calling this function may stop the timer but actually it enables this periph to be powermanaged either by the hardware pm or by software"

"*_stop" -> "*_pm_enable" or "*_pm_free" or "*_idle" 
"*_start" -> "*_pm_disable" or "*_pm_in_use" or "*_use" 

btw.: I also think the power management should be done by the periph driver but i am also thinking the function that does/enables the pm should reflect that they are mainly for pm and not other purposes (like stopping and starting)

the naming of these function may set precedent for other drivers 

this might as well have interface like `dev->pm(PM_FREE)` `dev->pm(PM_USE)`

Reply:
> if `*_stop` does may not stop the device and `*_start` may not start the device why have it named "start" and "stop" (we need to write "calling this function may stop the timer but actually it enables this periph to be powermanaged either by the hardware pm or by software"

I don't like the *may* here. It's the root cause of my problems. I want to actually stop the underlying hardware.

But yes, it could be extra cost. Especially in situations like:

```c
ztimer_stop(ZTIMER_USEC, &timeout_timer);
do_stuff();
ztimer_start(ZTIMER_USEC, &timeout_timer);
```

I try to gather some measurements how long `timer_start()` and `timer_stop()` take.

Reply:
> this might as well have interface like `dev->pm(PM_FREE)` `dev->pm(PM_USE)`

Yes, a common driver model would be very nice! But I don't want to wait for RIOT shifting to such a driver model ;-)

Reply:
Let's not overcomplicate this, and moreover, let's not stop having a ztimer_acquire in there on what can perfectly be a later optimization.

Reply:
#18821 points to a driver/device  that has the very problem that i am talking about the timer_stop in sam0 waits for the timer to actually stop (since it is called timer stop) which is not need for power_management and that device is one of the kind where waiting for sync takes ages in cpu time.

`timer_stop` is no power mangement function in that device (it might very well do PM ( hence #18821 is correct) but that is not its purpose. its purpose is to stop the timer and ensure it is stoped).

Reply:
I'll try to collect some facts by actually measuring the introduced delay.
Syncing two clock domains with fast clock should be pretty fast.

But even if significant delays are introduced: this driver can compensate for that if they are predictable.

Reply:
I wrote a little test program and executed on an `samr30-xpro` board on current master:

```c
    for (size_t i = 0; i < 10; i++) {
        // Just toggle the LED
        LED0_ON;
        LED0_OFF;
        // --> took 172ns

        // Timer stop
        LED0_ON;
        timer_stop(TIMER);
        LED0_OFF;
        // --> took 1668ns

        // Timer start
        LED0_ON;
        timer_start(TIMER);
        LED0_OFF;
        // --> took 1164ns
    }
```

![tek0006](https://user-images.githubusercontent.com/6105784/199257856-4013824e-8723-42cc-8d63-c39537b94f2d.png)

If I correct the timings by the LED toggle duration, I get the following durations:
- `timer_start()` took 992ns = 0.992us
- `timer_stop()` took 1496ns = 1.495us

I'd say this is acceptable, don't you think?

Reply:
We should not waste time waiting for things to turn off (or even worse off and then on again) when we want to save power.

off and on:
only with one msec user:
```
while(1){
do something; sleep_msec()
}
```
In every msec interrupt the timer will get turned off and for the next sleep it will be turned on again.


We should also not predictably waste time especially not in an interrupt (last ztimer of a clock gets handled (triggered) -> list is empty -> release gets called no user is left -> timer_stop() gets called -> time gets wasted while in IRQ if this happens for msec on a sam0 rtt mcu it needs to sync to the slow rtt  clock domain while in that IRQ)

Reply:
@jue89: on which timer was the test program?

Reply:
> @jue89: on which timer was the test program?

`periph_timer` with 1MHz. It will be the timer that's started and stopped regularly.

In every application, that's using `gnrc`, there's always a `ZTIMER_MSEC` running. Thus, `periph_rtt` isn't started and stopped during normal operation.

> We should also not predictably waste time especially not in an interrupt (last ztimer of a clock gets handled (triggered) -> list is empty -> release gets called no user is left -> timer_stop() gets called -> time gets wasted while in IRQ if this happens for msec on a sam0 rtt mcu it needs to sync to the slow rtt clock domain while in that IRQ)

But the time for house-keeping (i.e. calling the callback, changing the linked list, setting-up the next alarm) is way longer than the time consumed by `timer_start()` and `timer_stop()`. Especially on the `sam0` family an `periph_rtt`. I know the pain and setting up the next checkpoint is takes *many* cpu cycles.

On top, it's also possible to use `ztimer_periph_timer_ondemand` and skip `ztimer_periph_rtt_ondemand`. This will lead to a setup where just `timer_start()`/`timer_stop()` are called. `rtt_poweron()`/`rtt_poweroff()` are never called, even in the case there's no user left on the `ZTIMER_MSEC` clock.

--

I'm still convinced that this *the timer maybe stalls* isn't the right way to do it. And it's the only alternative I see to this approach. `ztimer` is not meant to be a high-precision timing source. If an application requires this, it should use a dedicated `periph_timer`.

Reply:
I did the same experiment on #18821 with the `samr30-xpro` board:
- `timer_start()`: 2.168us
- `timer_stop()`: 2.664us

-> the introduced `pm_block()`/`pm_unblock()` adds a delay of ~1.2us each! Most of the time is lost in `pm_layered`. And we can't skip interacting with it.

--

And the same experiment on #18780 (with #18814 applied):
- `timer_start()`: 2.500us
- `timer_stop()`: 2.580us

This also includes the interaction with `pm_layered`.

Reply:
Does this need to be sorted out necessarily for this PR to go in (vs. adjusting the implementation in a follow-up)? I don't understand all the discussion's details, but it appears to me that it's about performance of a weird use case (a loop repeatedly setting a timer would be better off setting a periodic timer and clearing it on a break), whereas this PR as a whole s necessary to make ztimer usable correctly.

Reply:
And the results of the test program executed on an weact-f411ce board on current master:
- `timer_stop()`: 0.264us
- `timer_start()`: 0.208us

I'm afraid, I don't have further boards at hand.